agent/engine/containermetadata/acquire_metadata.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+
+	"github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+// acquireNetworkMetadata parses the docker.NetworkSettings struct and
+// packages the desired metadata for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func acquireNetworkMetadata(settings *docker.NetworkSettings) NetworkMetadata {
+	if settings == nil {
+		seelog.Errorf("Failed to acquire network metadata: Network metadata not available or does not exist")
+		return NetworkMetadata{}
+	}
+
+	// This metadata is available in two different places in NetworkSettings
+	ipv4Address := settings.IPAddress
+	ipv4Gateway := settings.Gateway
+	ipv6Address := settings.GlobalIPv6Address
+	ipv6Gateway := settings.IPv6Gateway
+
+	// Network mode is not available for Docker API versions 1.17-1.20
+	networkModeFromContainer := ""
+	if len(settings.Networks) == 1 {
+		for modeFromSettings, containerNetwork := range settings.Networks {
+			networkModeFromContainer = modeFromSettings
+			ipv4Address = containerNetwork.IPAddress
+			ipv4Gateway = containerNetwork.Gateway
+			ipv6Address = containerNetwork.GlobalIPv6Address
+			ipv6Gateway = containerNetwork.IPv6Gateway
+		}
+	} else {
+		seelog.Debug("No network mode found due to old Docker Client version")
+	}
+
+	networkMD := NetworkMetadata{
+		networkMode: networkModeFromContainer,
+		ipv4Address: ipv4Address, ipv4Gateway: ipv4Gateway,
+		ipv6Address: ipv6Address,
+		ipv6Gateway: ipv6Gateway,
+	}
+	return networkMD
+}
+
+// acquireDockerContainerMetadata parses the metadata in a docker container
+// and packages this data for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func acquireDockerContainerMetadata(container *docker.Container) DockerContainerMD {
+	if container == nil {
+		seelog.Errorf("Failed to acquire container metadata: Container metadata not available or does not exist")
+		return DockerContainerMD{}
+	}
+
+	imageNameFromConfig := ""
+	// A real world container should never lack a config but we check regardless to avoid
+	// nil pointer exceptions (Could occur if there is some error in the docker api call)
+	if container.Config != nil {
```

Comment:
If I do not immediately return from the function, is this reverse conditional change still appropriate? I removed the `return DockerContainerMD{}` altogether because of the policy of continuing to parse the metadata even if a (vital) part of it is missing. 
Is there a way to make this more concise?
```
        imageNameFromConfig := ""
        if container.Config != nil {
                  imageNameFromConfig = container.Config.Image
          } else {
                  seelog.Warn("")
          }

```